packages/core/contracts/l2/polygon/tunnel/FxBaseRootTunnel.sol

```diff
@@ -0,0 +1,185 @@
+// SPDX-License-Identifier: MIT
```

Comment:
The architecture here is really interesting. Might be nice to have a README somewhere that explains how the state connection works. It seems that the architecture basically works this way:

- There's a root contract and a child contract.
- Contracts that want to communicate inherit from the root and child and specify one another as the root/child addresses.
- These inheriting contracts have some variant of "sendMessage" and "receiveMessage" to send and receive messages over the bridge.
- They call sendMessage to deliver bytes to the other side.
- They override a "process" method to actually receive and do stuff with the data received.

The way it seems to be set up, you would have each contract that wants to communicate replicate this same setup and effectively be a bridge in itself rather than sending messages to and from a bridge. This is a really convenient architecture for permissioning purposes (despite being a bit bloated) since contracts are initialized in pairs that have a very obvious permissioning structure (they only allow messages to and from each other).

Assuming this is all correct, I'm still wanting to figure out a way where we don't have to couple our implementations of the bridged contracts (like the oracle and governor bridges) to whatever type of bridge they're interacting with.

We could try to make all of our bridge implementations look like this sort of inheritance tree.

We have an abstract oracle and governor contracts that define an unimplemented _sendMessage(bytes) function that needs to be implemented by a derived class. They also implement a _receiveMessage(bytes) function.

Then there are bridge adapter contracts that you can inherit from. They are also abstract. They implement a _sendMessage(bytes) function, but they have an unimplemented _receiveMessage(bytes) function that needs to be implemented by a derived class. Basically the opposite of the above.

Then all we would (theoretically need to do) is have a derived contracts that look like this (literally no code other than combining base classes). Note that this is not tested, so I haven't verified that this type of inheritance actually fills in the missing abstract implementations as one would expect.:
```
contract PolygonOracleSink is PolygonChildBridge, OracleSink {}

contract PolygonGovernorSink is PolygonChildBridge, GovernorSink {}

contract OptimismOracleSink is OptimismChildBridge, OracleSink {}

contract OptimismGovernorSink is OptimismChildBridge, GovernorSink {}

...
```

Thoughts? Just trying to brainstorm ways to generalize a bit so the only thing we need to add on each new bridge is the bridging logic, itself.

Reply:
I'll add a readme and re-ping you for review!

Thinking about this proposed architecture.

Reply:
Added read me with diagram